tests/training/optimizer_test.py

```diff
@@ -77,6 +77,116 @@ def test_optimizer_parameter_groups(self):
+                # as well as any parameters which didn't match one of the given regexes.
+                assert len(param_groups) == 2
+
+    def test_optimizer_params(self):
+        optimizer_params = Params({
+                                    "type": "regex",
+                                    "optimizers": [
+                                        {"name": "default", "type": "adam", "lr": 1},
+                                        {"name": "embedder", "type": "adam", "lr": 2},
+                                        {"name": "encoder", "type": "adam", "lr": 3},        
+                                    ],
+                                    "parameter_groups": [
+                                        [["^text_field_embedder"], {"name": "embedder", "betas": (0.9, 0.98), "lr": 2, "weight_decay": 0.01}],
+                                        [["^encoder.*bias"], {"name": "encoder", "lr": 0.001}],
+                                        [["^encoder.*weight"], {"name": "encoder", "lr": 0.002}],
+                                        [["^tag_projection_layer.*.weight$"], {"lr": 5}]
+                                    ],
+                                    })
+        import torch
+       
+        parameters = [[n, p] for n, p in self.model.named_parameters() if p.requires_grad]
+        optimizer = Optimizer.from_params(model_parameters=parameters, params=optimizer_params)
+
+        # When the RegexOptimizer is initialized, `optimizer.param_groups` stores the parameter groups.
+        # These parameter groups are assigned to their own optimizer by the RegexOptimizer.
+        # Check that changes to the parameters in optimizer.param_groups affect the parameters in optimizer._grouped_optimizers.
+        regex_optimizer_params = []
+        regex_optimizer_grouped_optimizer_params = []
+
+        for param_group in optimizer.param_groups:
+            # Each param_group should have optimizer options visible so they can be used by schedulers.
+            lr = param_group["lr"]
+            assert lr > 0
+            params = param_group["params"]
+            for param in params:
+                param.data.zero_()
+                regex_optimizer_params.append(id(param))
+
+        # Check that the parameters of the sub-optimizers were also changed.
+        for optimizer in optimizer._grouped_optimizers.values():
+            for param_group in optimizer.param_groups:
+                params = param_group["params"]
+                for param in params:
+                    regex_optimizer_grouped_optimizer_params.append(id(param))
+                    assert param.sum() == 0, "Param has non-zero values."
+
+        # As the optimizers are created in accordance with the order of groups in `parameter_groups`, the order of parameters should be deterministic.
+        assert regex_optimizer_params == regex_optimizer_grouped_optimizer_params
+
+
 class TestDenseSparseAdam(AllenNlpTestCase):
```

Comment:
There is a test failing here for a device type. It is not related to the `RegexOptimizer` tests but I didn't manage to fix it.

Reply:
Don't worry about it if it's clearly not related. I'll take a look if it comes back.